src/mame/drivers/rastersp.cpp

```diff
@@ -436,6 +629,193 @@ WRITE_LINE_MEMBER( rastersp_state::vblank_irq )
 		update_irq(IRQ_VBLANK, ASSERT_LINE);
 }
 
+/*
+	Mimic a Bacta Dataport unit just enough to satisfy Football Crazy.
+	Also mimic trackball unit. This uses a small PIC based interface board
+	which converts the trackball movement into RS232 data.
+	The data format is as follows:
+	Byte 0 : 0100abAB bits 7 & 6 = 01 signifies 1st byte
+	Byte 1 : 00CDEFGH bits 7 & 6 = 00 signifies 2 or 3rd byte (not identified)
+	Byte 2 : 00cdefgh
+
+	The trackball X movement is ABCDEFGH and Y movement abcdefgh
+	We do not have a PCB nor a dump of the PIC :(
+*/
+TIMER_CALLBACK_MEMBER(fbcrazy_state::edc_tb_timer)
+{
+	uint8_t	newxpos, newypos;
+
+	// Check if uart interrupts are enabled
+	if (m_uart_regs[0][9] != 0)
+	{
+		// Check for pending tx data
+		if (m_uart_ctrl != 0)
+		{
+			// Set tx complete!
+			m_uart_ints |= m_uart_ctrl;
+			m_uart_ctrl = 0;
+			update_irq(IRQ_UART, 1);
+		}
+		if (m_uart_tx_count[0] >= 2)
+		{
+			// At least 2 bytes have been received - most edc packets are 2 bytes
+			// so we can just send ack to keep Rasterspeed system happy
+			// and prevent it from locking up
+			m_uart_rx_data[0][1]=0x06;//ack
+			m_uart_rx_data[0][0]=0x01;//1 byte to send
+			m_uart_ints |= 0x02;//flag rx int
+			m_uart_tx_count[0]=0;
+			update_irq(IRQ_UART, 1);
+		}
+
+		// Check rx ints are enabled for trackball
+		if ((m_uart_regs[1][1] & 0x18) == 0x10 && (++m_trackball_ctr > 3) && (m_uart_ints & 0x08) == 0)
+		{
+			m_trackball_ctr = 0;
+
+			newxpos = ioport("TRACK_X")->read();
+			newypos = ioport("TRACK_Y")->read();
+
+			if (newxpos != m_oldxpos || newypos != m_oldypos)
+			{
+				int diffx, diffy;
+				diffx = newxpos - m_oldxpos;
+				diffy = newypos - m_oldypos;
+				m_oldxpos = newxpos;
+				m_oldypos = newypos;
+
+				//Format movement into 3 byte as described above and tx
+				m_uart_rx_data[1][3] = 0x40 | ((diffy>>4)&0x0c) | ((diffx>>6)&0x03);
+				m_uart_rx_data[1][2] = diffx & 0x3f;
+				m_uart_rx_data[1][1] = diffy & 0x3f;
+				m_uart_rx_data[1][0] = 3;
+				m_uart_ints |= 0x08;
+				update_irq(IRQ_UART, 1);
+			}
+		}
+	}
+}
+
+uint8_t fbcrazy_state::edc_tb_r(offs_t offset)
```

Comment:
What UART is this supposed to be?  The simulation looks complex enough that it might justify a device.

Reply:
It's a 85C30. I know that this is already supported however, when I originally looked at using it with an external clock I mistook the comment "Support for external clock as source for BRG yet to be finished" in the file z80scc.cpp to mean that mode of operation wasn't supported and I couldn't understand why and what to do with it. Looking again the comment is a bit misleading and it does support an external clock - just nothing to do with that function!
If I try and update to this I should be able to use the BACTA datalogger mentioned above. That just leaves the track ball interface which is serial. I'd have to make a serial device for that to connect to the 85c30. We don't have the code for the track ball interface board so can't emulate it.
Is that what you'd prefer ? So remove all the stuff that's in there at the moment and replace with 85c30, datalogger and track balll device ?

Reply:
Yeah, better to use the 85C30 device.  It’s known to be buggy, but more things using it will hopefully help get more bugs fixed.  You don’t necessarily need to make a device for the trackball – you could hook up callbacks in the driver state class to the SCC’s transmit/receive lines and simulate the trackball there.  If you do decide to create a device, the `device_serial_interface` mixin should simplify the implementation if it isn’t doing anything exotic.

Reply:
Trackball is just a tx device, no handshake or anything so just need to worry about tx line. But I'd still need to create a stream of serial data if I hook onto the scc receive line, unless I misunderstood what you meant ?  Looking at z80scc.cpp shows that receive_data(data) which places an rx'd byte into the fifo is public so could use that and push straight into the fifo ? Or would that be frowned upon.